### PR TITLE
ENH: Added QR solver using QR decomposition

### PR DESCRIPTION
**Pull request by Tsangares**
This is a method for solving linear systems using QR factorization and back substitution. I found no solvers like this in numpy or scipy so when I build it I thought it might help others to have. 

#### Reference issue
No solver if the matrix is m by n 

#### What does this implement/fix?
It adds a QR decomposition solver using back substitution.

#### Additional information
I am still going through the contribution docs to make sure everything is okay. Let me know if I did something wrong. My commit message was wrong; it did not have a prefix. I hope that is okay.

### CODE REVIEW

**Review by dschmitz89 — COMMENTED** [review]
Thank you for the PR!

How does this compare to Lapack's solvers? From what I see, [dgels](https://netlib.org/lapack/explore-html/d7/d3b/group__double_g_esolve_ga225c8efde208eaf246882df48e590eac.html) also performes a QR decomposition under the hood.

**Comment by ilayn** [comment]
Hi @Tsangares thanks for the PR. If you use `scipy.linalg.lstsq` with driver `gelsy`, `gels` it does what you are computing here for the general case. There are multiple steps you have that will not lead to an acceptable solution namely, first using a matrix inverse which is a no-go (it should use at least `linalg.solve` but then you don't need QR composition) and also matrix_rank which uses an SVD internally hence increasing complexity way above acceptable levels from a solver. Catching the correct number of zero rows is also a very nontrivial step which these low level routines provide already. 

Alternatively if you want to use specifically QR factorization then you can use `geqrf` and work from there via calling `trtrs`. See https://stackoverflow.com/questions/21970510/solving-a-linear-system-with-lapacks-dgeqrf for an example. In summary, we also provide these low level LAPACK functions for individuals to cook up their own chain of linear algebra steps in an efficient manner. See https://docs.scipy.org/doc/scipy/reference/linalg.lapack.html for the list and let us know if you are missing any.

**Comment by Tsangares** [comment]
Thank you for your quick reply. I realize I titled this as a solution to non-square matrices, but I wrote this code originally when solving a system where all the coefficients are integers. All the `lstsq` methods seem to give a numerical result through optimization rather than an algebraic one. Here is what one of the a `lstsq` solutions looks like as an example using, `gelsd`, `gelsy`, `gelss` or `dgels`:

gelss

Here is using the QR-Decomp solver:

qr_decomp

The QR-decomp find an algebraic or close solution given `A` is `nxm`. You can see from the last plot the solution at the top does not use all the rows in the matrix and results in integer coefficients. This allows me to further decompose the system with analysis. I do use the inbuilt QR decomp package, `scipy.linalg.qr`, from scipy which should be fast. Should I switch to implementing the LAPACK `DORMQR`?

I switched the `np.linalg.inv` for the `linalg.solve`. I am looking how to get the permutation matrix. I might be able to find the rank from the QR decomp, would that be sufficient or is this still not quick enough solver?

**Comment by josef-pkt** [comment]
IIUC

The lapack functions like gelsy compute a minimum norm solution (similar to using pinv)
This solution is not "sparse" setting redundant parameters to zero.

IIUC, this is then the same or similar problem that we have with rank deficient design matrices in linear regression. 
statsmodels and sklearn use pinv, Stata and R drop columns using some pivoting QR, i.e. set those coefficients to zero.

I guess it's only relevant in underdetermined systems, otherwise the LS solution is unique.

**Comment by Tsangares** [comment]
@josef-pkt I switched it for `scipy.linalg.solve` in #51e7bcded3851df7c2abf4366495ab746efc00b4

**Comment by ilayn** [comment]
You can certainly do what you are doing but this is not a generic solver to be included in SciPy for reasons I mentioned above. One is that it is not really a solver but a script that performs factorization and rank selection and solving via the regular gesv hence discarding the structure of R and orthogonalilty of Q. 

What you have in mind goes by the name rank-revealing QR factorizations. A standard QR factorization won't cut it. And in fact it is highly nontrivial. Note that it should work for all cases and not some integer entry matrices to be included in SciPy namespace. 

>  through optimization rather than an algebraic one

It is still algebraic but comes with the additional property of being the minimum norm solution.

**Comment by josef-pkt** [comment]
@ilayn Isn't the pivoting QR rank revealing with a cutoff on the diagonal elements of R?

I was playing with this recently, but didn't like much which columns were set to zero, i.e. the pivoting, permutation result. I have no intuition what would be dropped in this case.

(statsmodels will eventually be forced to add something like this for OLS because R and Stata users want it. Currently still undecided how to add it.)

**Comment by ilayn** [comment]
No column pivoting QR (such DGEQP3 in LAPACK) is an alternative to rank revealing QR but they are not still water tight. I don't quite understand what the use case is could you cook up something? Why QR specifically and not LU or SVD?

**Comment by josef-pkt** [comment]
OLS in statsmodels and linearregression in sklearn use svd/pinv. However that gives us the minimum norm solution if the system is underdetermined.
Most users would prefer setting some parameters to zero instead of the minimum norm solution.

Our case for this is: `A` matrix has more rows than columns, so would be overdetermined (LS solution) if column rank is full.
The perfect collinearity case is if some columns are linear combinations, i.e. no full column rank.

(I can try to find the examples that I used for pivoting QR, after cooking and dinner)

File: scipy/linalg/_qr_solve.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ def qr_solve(A,y,silent=True):
     R_2 = R[:,:-(rank-m)]
     
     #z_1 is a column vector with r elements
-    z_1 = np.linalg.inv(R_1)@Q_1.T@y
+    z_1 = scipy.linalg.solve(R_1,Q_1.T@y)
```
Comment (Tsangares) [inline scipy/linalg/_qr_solve.py:79]: I removed the inverse and changed it to `scipy.linalg.solve()`

**Comment by josef-pkt** [comment]
here's a baby example that I had used to look at pivoting qr

```
from scipy import linalg
a = np.random.randn(10, 4)
a[:, 2] = a[:, :2].sum(1)
a[:, 3] = a[:, :2].sum(1) + 1e-10 * a[:, 3]
q, r, p = linalg.qr(a, mode='economic', pivoting=True)
q, r, p

np.diag(r)
array([ 4.94094427e+00, -1.73703024e+00, -2.76238683e-10,  9.34178911e-16])
```

**Comment by Tsangares** [comment]
@josef-pkt What are you looking for? Maybe I can help.

**Comment by josef-pkt** [comment]
@Tsangares  decision for statsmodels in how to add this and which version is still open.
We have a similar PR https://github.com/statsmodels/statsmodels/pull/6935 
There are still many open design decisions how to add this to statsmodels, and, personally, I prefer the pinv minimum norm solution.
So I have not pushed for it yet, in spite of the frequent user requests.

However the code here is useful to us, except that I would base the cutoff based on diag(r) instead of svd matrix rank.
(We need more than just the solution of the QR solve, because we also need standard errors. So, statsmodels is not directly using the scipy or numpy linalg functions that throw away some of the needed intermediate computations. e.g. our pinv used in ols and other linear models keeps also singular values  https://github.com/statsmodels/statsmodels/blob/main/statsmodels/tools/tools.py#L255)

**Comment by ilayn** [comment]
The sparse solution is also very desirable for compressed sensing and for some other fields. But by the way finding the sparsest solution for underdetermined systems is NP-hard. 

Folks use basis pursuits and other techniques that I'm not very familiar with through L1 optimization instead of L2 like SVD based methods give (actually the problem is an L0 problem). I haven't checked the literature for that for some time now. But I think randomized QR pivoting is the hot take. 

What I do remember is that column pivoting is reported as much less reliable than SVD and cost-wise it is not that far off from SVD. In fact, if you have integer systems I'd even consider solving them in a symbolic fashion with exact arithmetic.

But obviously, if you need it, you should do it. For SciPy I think a regular QR -> truncate -> solve_triangular + @ Q.T is a valid solution. I think it doesn't warrant for a complete new function in the public API because it is not general enough in my opinion (loosely held). And also truncation tolerance is tricky so if you would leave it to user then this is pretty much a custom script.

File: scipy/linalg/_qr_solve.py [hunk 2]
```diff
@@ -63,7 +63,7 @@ def qr_solve(A,y,silent=True):
     P = np.eye(len(P))[:,P]
     
     #Let r be the number of linearly independent columns & rows in A (the rank)
-    rank = r = np.linalg.matrix_rank(A)
+    rank = r = sum(np.around(np.diag(R),12)!=0)
```
Comment (Tsangares) [inline scipy/linalg/_qr_solve.py:66]: @josef-pkt Was right, I can obtain the rank from using the diagonal of R.

Comment (Tsangares) [reply scipy/linalg/_qr_solve.py:66]: set the tolerance for zero to be `e-12`.

**Comment by Tsangares** [comment]
@josef-pkt I see you dislike how the permutation matrix dictates the solution as there are other solutions. I have been trying to find a solution set in [this gist here ](https://gist.github.com/Tsangares/fddc14e40a2b4e285db5aaac07226acf)but to no particular avail. Although, its a nice template if you want to mess with this QR. 

What is this pinv minimal norm? 

@ilayn I remove the `inv` and the call to `matrix_rank` so its just using QR decomposition with a pivot. 

I see what you mean about this method being a specific method. Although, this is not the first time I have come across needing a solution like this and matlab and other platforms do have it. 

I am happy to help though to find an optimal solution (no pun intended).

**Comment by ilayn** [comment]
I am still trying to understand what the property of this solution is. Or what it is guaranteeing if compared to `lstsq`. 

> matlab and other platforms do have it.

Which command in matlab do you have in mind? Maybe that would clarify more things for me. Note that matlab move away from geqrf to geqrt very recently hence things might have changed.

**Comment by Tsangares** [comment]
> I am still trying to understand what the property of this solution is. Or what it is guaranteeing if compared to `lstsq`.
> 
> > matlab and other platforms do have it.
> 
> Which command in matlab do you have in mind? Maybe that would clarify more things for me. Note that matlab move away from geqrf to geqrt very recently hence things might have changed.

I am thinking `rref` then using the pivot to find a solution in a quotient space: 
```
function B = rrefmodp(A, p)

%rrefmodp -- reduced row echelon form (mod p)
%   B = rrefmodp(A,p)
%
%[VdS 2012-aug-2]

%% compile the inverse look-up table.
I = zeros(1,p-1);
Ma = zeros(1,p-1);
for a = (1: p-1)
    Ma = mod(Ma+(1:p-1), p);
    I(a) = find(Ma == 1);
end

%% normalise A to take values in (0:p-1)
A = int32(A);
A = mod(A,p);

[m,n] = size(A);

%% main loop

% set pointers to next pivot
a = 1;
b = 1;

% row permutation
P = (1:m);

while ((a <= m) & (b <= n))
    % if next entry is zero...
    if (A(P(a),b) == 0)
        nzb = find(A(P(a+1:end),b));

       % if column is empty, advance column
       if isempty(nzb)
           b=b+1;
           continue
       end
       
       % else switch rows
       
       a2 = a + nzb(1);
       P([a a2]) = P([a2 a]);
    end
    
    % normalise row
    A(P(a), (b+1:end)) = mod(A(P(a), (b+1:end)) * I(A(P(a),b)), p);
    A(P(a), b) = 1;
    
    % clear remaining rows
    nzb_log = logical(A(:,b));
    nzb_log(P(a)) = 0;
    
    nzb = find(nzb_log);
    for c = 1: length(nzb)
        A(nzb(c), (b+1: end)) = A(nzb(c), (b+1: end)) ...
                                - (A(nzb(c),b) * A(P(a), (b+1:end)));
        A(nzb(c), (b+1: end)) = mod( A(nzb(c), (b+1: end)), p);
    end
    A(nzb_log, b) = 0;

    a = a + 1;
    b = b + 1;
end

B = A(P,:);

return
```

**Comment by j-bowhay** [comment]
Thanks for the pr @Tsangares! I am think consensus here is we probably won't move forward with this but thanks for your contribution regardless

**Comment by Tsangares** [comment]
> Thanks for the pr @Tsangares! I am think consensus here is we probably won't move forward with this but thanks for your contribution regardless

Hey, what did you all decide on the solver? Or is it still pending?

**Pull request closed without merging**